File: packages/blognami/lib/models/pageable.js

import { inflector } from 'haberdash';

export default {
    meta(){
        this.beforeValidation(async function(){
            const slug = `${this.slug || ''}`.trim();
            if(slug == ''){
                let n = 1;
                while(true){
                    const candidateSlug = this.generateCandidateSlug(n);
                    if(!(await this.database.pageables.where({ slug: candidateSlug }).first())){
                        this.slug = candidateSlug;
                        break;
                    }
                    n++;
                }
            }
        });

        this.validateWith(async pageable => {
            if(!pageable.isValidationError('slug') && await pageable.database.pageables.where({ idNe: pageable.id, slug: pageable.slug }).count()){
                pageable.setValidationError('slug', 'Must be unique');
            }
        });
    },

    generateCandidateSlug(n){
        const out = inflector.dasherize(this.title || this.name);
        if(n > 1){
            return `${out}-${n}`;
        }
        return out;
    }
};

File: projects/blognami/lib/blognami/models/pageable.rb

require "securerandom"

require "slick/database/row"



Slick::Database::Row.register "pageable", abstract: true do
  
    before_validation do
        next if !slug.nil?
        n = 0
        while true
            candidate_slug = generate_candidate_slug(n)
            if database.pageables.slug_eq(candidate_slug).count == 0
                self.slug = candidate_slug
                break
            end
            n += 1
        end
    end

    def generate_candidate_slug(n)
        out = if respond_to? :title
            title.dasherize
        elsif respond_to? :name
            name.dasherize
        else
            SecureRandom.uuid
        end
        return "#{out}-#{n}" if n > 0
        out
    end

end
